src/CoreWCF.Primitives/src/CoreWCF/Security/SecuritySessionSecurityTokenAuthenticator.cs

```diff
@@ -0,0 +1,1366 @@
+            // securityChannelListener.InitializeListener(channelBuilder);
+            this.shouldMatchRstWithEndpointFilter = SecurityUtils.ShouldMatchRstWithEndpointFilter(this.bootstrapSecurityBindingElement);
+            //return securityChannelListener;
+        }
+
+       internal class SecuritySessionHost //: ServiceHostBase
+        {
+            //ChannelBuilder channelBuilder;
+            MessageFilter filter;
+            Uri listenUri;
+            SecuritySessionSecurityTokenAuthenticator authenticator;
+
+            public SecuritySessionHost(SecuritySessionSecurityTokenAuthenticator authenticator, MessageFilter filter, Uri listenUri)//, ChannelBuilder channelBuilder)
+            {
+                this.authenticator = authenticator;
+                this.filter = filter;
+                this.listenUri = listenUri;
+               // this.channelBuilder = channelBuilder;
+            }
+
+            //protected override ServiceDescription CreateDescription(out IDictionary<string, ContractDescription> implementedContracts)
+            //{
+            //    implementedContracts = null;
+            //    return null;
+            //}
+
+            
+            internal ChannelDispatcher InitializeRuntime(SecurityServiceDispatcher securityDispatcher)
+            {
+                MessageFilter contractFilter = this.filter;
+                int filterPriority = Int32.MaxValue - 10;
+                List<Type> endpointChannelTypes = new List<Type> {  typeof(IReplyChannel),
+                                                           typeof(IDuplexChannel),
+                                                           typeof(IReplySessionChannel),
+                                                           typeof(IDuplexSessionChannel) };
+
+                //  IChannelListener listener = null;
+                //  BindingParameterCollection parameters = new BindingParameterCollection(this.channelBuilder.BindingParameters);
+                //  Binding binding = this.channelBuilder.Binding;
+                //  binding.ReceiveTimeout = this.authenticator.NegotiationTimeout;
+                // parameters.Add(new ChannelDemuxerFilter(contractFilter, filterPriority)); 
+                // DispatcherBuilder.MaybeCreateListener(true, endpointChannelTypes, binding, parameters,
+                //                                      this.listenUri, "", ListenUriMode.Explicit, this.ServiceThrottle, out listener);
+                //   if (listener == null)
+                // {
+                //      throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.CannotCreateTwoWayListenerForNegotiation)));
+                //  }
+
+                //Replacing above code by below 3 lines.(adding securityservicedispatcher to demuxer, how to respond)
+                Binding binding = this.authenticator.IssuerBindingContext.Binding;
```

Comment:
You probably want to clone the binding as you then need to modify the receive timeout to be this.authenticator.NegotiationTimeout and you never modify a copied Binding without cloning it first.